lib/Dialect/SystemC/Transforms/SystemCLowerInstanceInterop.cpp

```diff
@@ -0,0 +1,186 @@
+    OpBuilder includeBuilder(hwModule);
+    includeBuilder.create<emitc::IncludeOp>(loc, tn + ".h", false);
+
+    // Request a pointer to the verilated module as persistent state.
+    Value state = rewriter
+                      .create<interop::ProceduralAllocOp>(loc, stateType,
+                                                          InteropMechanism::CPP)
+                      .getStates()[0];
+
+    insertStateInitialization(rewriter, loc, state);
+
+    ValueRange results = insertUpdateLogic(
+        rewriter, loc, state, adaptor.getInputs(), op.getResults(),
+        adaptor.getInputNames(), adaptor.getResultNames());
+
+    insertStateDeallocation(rewriter, loc, state);
+
+    // Replace the return values of the instance with the result values of the
+    // interop update operation.
+    rewriter.replaceOp(op, results);
+    return success();
+  }
+
+private:
+  /// Insert a interop init operation to allocate an instance of the verilated
+  /// module on the heap and let the above requested pointer point to it.
+  void insertStateInitialization(PatternRewriter &rewriter, Location loc,
+                                 Value state) const {
+    auto initOp = rewriter.create<interop::ProceduralInitOp>(
+        loc, state, InteropMechanism::CPP);
+
+    OpBuilder initBuilder = OpBuilder::atBlockBegin(initOp.getBody());
+    Value newState =
+        initBuilder.create<NewOp>(loc, state.getType(), ValueRange());
+    initBuilder.create<interop::ReturnOp>(loc, newState);
+  }
+
+  /// Create an update interop operation to assign the input values to the input
+  /// ports of the verilated module, call 'eval', and read the output ports of
+  /// the verilated module.
+  ValueRange insertUpdateLogic(PatternRewriter &rewriter, Location loc,
+                               Value stateValue, ValueRange inputValues,
+                               ValueRange resultValues, ArrayAttr inputNames,
+                               ArrayAttr resultNames) const {
+    auto updateOp = rewriter.create<interop::ProceduralUpdateOp>(
+        loc, resultValues.getTypes(), inputValues, stateValue,
+        InteropMechanism::CPP);
+
+    OpBuilder updateBuilder = OpBuilder::atBlockBegin(updateOp.getBody());
+
+    // Write to the verilated module's input ports.
+    Value state = updateOp.getBody()->getArguments().front();
+    for (size_t i = 0; i < inputValues.size(); ++i) {
+      Value member = updateBuilder.create<MemberAccessOp>(
+          loc, inputValues[i].getType(), state,
+          inputNames[i].cast<StringAttr>(), MemberAccessKind::Arrow);
+      updateBuilder.create<AssignOp>(loc, member,
+                                     updateOp.getBody()->getArgument(i + 1));
+    }
+
+    // Call 'eval'.
+    auto evalFunc = updateBuilder.create<MemberAccessOp>(
+        loc, FunctionType::get(updateBuilder.getContext(), {}, {}), state,
+        "eval", MemberAccessKind::Arrow);
```

Comment:
Raw strings always seem weird. However, this "eval" is only ever used once. It is likely fine to not hoist this into a header.